File: solutions/lua/circular-buffer/circular-buffer.lua
local CircularBuffer = {}
CircularBuffer.__index = CircularBuffer

function CircularBuffer:new(size)
    local self =
        setmetatable(
        {
            capacity = size,
            values = {}
        },
        CircularBuffer
    )
    self:clear()
    return self
end

function CircularBuffer:read()
    if self.count == 0 then
        error("buffer is empty")
    end
    result = self.values[self.head]
    self.head = (self.head + 1) % self.capacity
    self.count = self.count - 1
    return result
end

function CircularBuffer:write(value)
    if value == nil then
        return
    end
    if self.count == self.capacity then
        error("buffer is full")
    end
    self:forceWrite(value)
end

function CircularBuffer:forceWrite(value)
    self.values[self.tail] = value
    self.tail = (self.tail + 1) % self.capacity
    if self.count == self.capacity then
        self.head = self.tail
    else
        self.count = self.count + 1
    end
end

function CircularBuffer:clear()
    self.count = 0
    self.head = 0
    self.tail = 0
end

return CircularBuffer

File: solutions/javascript/circular-buffer/circular-buffer.js
class CircularBuffer {
  #head = -1;
  #tail = -1;
  #size = 0;
  #values;

  constructor(size) {
    this.#size = size;
    this.#values = new Array(size);
  }

  write(value) {
    if (this.#isFull()) {
      throw new BufferFullError();
    }
    if (this.#isEmpty()) {
      this.#tail = this.#head = 0;
    } else {
      this.#tail = (this.#tail + 1) % this.#size;
    }
    this.#values[this.#tail] = value;
  }

  read() {
    if (this.#isEmpty()) {
      throw new BufferEmptyError();
    }
    let result = this.#values[this.#head];
    this.#values[this.#head] = undefined;
    if (this.#head == this.#tail) {
      this.clear(); // final element was removed from buffer
    } else {
      this.#head = (this.#head + 1) % this.#size;
    }
    return result;
  }

  forceWrite(value) {
    if (this.#isFull()) {
      this.#tail = this.#head;
      this.#head = (this.#tail + 1) % this.#size;
      this.#values[this.#tail] = value;
    } else {
      this.write(value);
    }
  }

  clear() {
    this.#head = this.#tail = -1;
    this.#values.fill(undefined);
  }

  #isEmpty() {
    return this.#head < 0;
  }

  #isFull() {
    return this.#head === (this.#tail + 1) % this.#size;
  }
}

export default CircularBuffer;

export class BufferFullError extends Error {
  constructor() {
    super("BufferError: Circular buffer is full");
  }
}

export class BufferEmptyError extends Error {
  constructor() {
    super("BufferError: Circular buffer is empty");
  }
}
